056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,129 @@
+and [PR #8892](https://github.com/strimzi/strimzi-kafka-operator/pull/8892).
+
+There is a related proposal describing how to diff controller configuration currently being discussed: [PR #82](https://github.com/strimzi/proposals/pull/82)
+
+## Current situation
+
+When operating on a ZooKeeper based cluster the KafkaRoller has the behaviour 
+described below.
+
+### Order to roll nodes
+Roll in order:
+ - Unready brokers
+ - Ready brokers
+ - Current controller
+
+### Triggers
+The following triggers an attempt to roll a broker:
+- Pod is annotated for manual rolling update
+- Pod is unready
+- Broker's configuration has changed and cannot be updated dynamically
+
+### Rollability checks
+ - Attempts to connect an admin client to the broker and if it can't connect rolls the broker
+ - Does not force roll a broker performing log recovery
+ - Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+### Configuration changes
+ - Compares the current configuration with expected configuration to see if a broker needs rolling
+ - Performs dynamic configuration updates if possible, otherwise rolls broker on configuration change
+
+In KRaft mode the KafkaRoller currently skips controller only nodes, but performs the above steps on any combined or broker only nodes.
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will wait still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When moving to KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+- Unready controller/combined nodes
+- Ready controller/combined nodes
+- Active controller
+- Unready broker-only nodes
+- Ready broker-only nodes
+
+If KafkaRoller fails to roll the controller/combined nodes, it still tries to roll the brokers before reporting an error.
+This is required because until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) 
+is implemented in order to check the quorum state when rolling controllers at least one broker needs to be running.
+
+### Triggers
+The following will trigger an attempt to roll a KRaft controller or combined node:
+- Pod is annotated for manual rolling update
+- Pod is unready
```

Comment:
Should we still list a configuration change here as a trigger? Since we do expect that to result in pods rolling, it just might be implemented in different ways depending on whether the KIP is complete